### PR TITLE
New: Kit-Library - Favorites

### PR DESCRIPTION
**Pull request by Nevoss**
ref: ED-2598

### CODE REVIEW

File: core/app/modules/kit-library/data/repository.php [hunk 1]
```diff
@@ -2,12 +2,15 @@
 namespace Elementor\Core\App\Modules\KitLibrary\Data;
 
 use Elementor\Core\Utils\Collection;
+use Elementor\Core\App\Modules\KitLibrary\Data\Exceptions\Kit_Not_Found_Exception;
```
Comment (StyleShit) [inline core/app/modules/kit-library/data/repository.php:5]:
```suggestion
use Elementor\Core\App\Modules\KitLibrary\Data\Exceptions\Kit_Not_Found_Exception;
use Elementor\Core\Utils\Collection;
```

Comment (StyleShit) [reply core/app/modules/kit-library/data/repository.php:5]: Ordered alphabetically.

File: core/app/modules/kit-library/data/repository.php [hunk 2]
```diff
@@ -86,6 +103,62 @@ public function get_taxonomies( $force_api_request = false ) {
...
+	/**
+	 * @param $user_id
+	 * @param $id
+	 *
+	 * @return array
+	 * @throws Exceptions\Api_Response_Exception
+	 * @throws Exceptions\Api_Wp_Error_Exception
+	 * @throws Kit_Not_Found_Exception
+	 */
+	public function add_to_favorite( $user_id, $id ) {
```
Comment (StyleShit) [inline core/app/modules/kit-library/data/repository.php:115]:
```suggestion
	public function add_to_favorites( $user_id, $id ) {
```

File: core/app/modules/kit-library/data/repository.php [hunk 3]
```diff
@@ -86,6 +103,62 @@ public function get_taxonomies( $force_api_request = false ) {
...
+	/**
+	 * @param $user_id
+	 * @param $id
+	 *
+	 * @return array
+	 * @throws Exceptions\Api_Response_Exception
+	 * @throws Exceptions\Api_Wp_Error_Exception
+	 * @throws Kit_Not_Found_Exception
+	 */
+	public function remove_from_favorite( $user_id, $id ) {
```
Comment (StyleShit) [inline core/app/modules/kit-library/data/repository.php:142]:
```suggestion
	public function remove_from_favorites( $user_id, $id ) {
```

**Review by StyleShit — CHANGES_REQUESTED** [review]
Nice work! 😄

**Review by StyleShit — APPROVED** [review]
LGTM

File: core/app/modules/kit-library/assets/js/hooks/use-kit-favorites-mutations.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+					if ( ! kits ) {
+						return kits;
+					}
+
+					return kits.map( ( item ) => item.id === kit.id ? kit : item );
+				}
+			);
+		}
+
+		queryClient.invalidateQueries( [ kitKey, data.data.id ] );
```
Comment (matipojo) [inline core/app/modules/kit-library/assets/js/hooks/use-kit-favorites-mutations.js:26]: Explain

File: core/app/modules/kit-library/assets/js/pages/index/index-sidebar.js [hunk 5]
```diff
@@ -1,13 +1,23 @@
...
+				className={ `eps-menu-item__link ${ ! props.filter.favorite ? 'eps-menu-item--active' : '' }` }
 				icon="eicon-filter"
-				url="/kit-library"
+				onClick={ () => props.filter.favorite && props.onChange( { ...initialFilterState } ) }
+			/>
+			<MenuItem
+				text={ __( 'Favorites', 'elementor' ) }
+				className={ `eps-menu-item__link ${ props.filter.favorite ? 'eps-menu-item--active' : '' }` }
+				icon="eicon-heart-o"
+				onClick={ () => ! props.filter.favorite && props.onChange( {
```
Comment (matipojo) [inline core/app/modules/kit-library/assets/js/pages/index/index-sidebar.js:17]: Use a route + current route

File: core/app/modules/kit-library/assets/js/pages/index/index.js [hunk 6]
```diff
@@ -106,7 +108,10 @@ export default function Index() {
 						{ isLoading && __( 'Loading...', 'elementor' ) }
 						{ isError && __( 'An error occurred', 'elementor' ) }
 						{ isSuccess && data.length > 0 && <KitList data={ data }/> }
-						{ isSuccess && data.length <= 0 && <IndexNoResults /> }
+						{
+							isSuccess && data.length <= 0 &&
```
Comment (matipojo) [inline core/app/modules/kit-library/assets/js/pages/index/index.js:112]:
```suggestion
							isSuccess && ! data.length &&
```

File: core/app/modules/kit-library/assets/js/pages/overview/overview.js [hunk 7]
```diff
@@ -41,7 +41,7 @@ export default function Item( props ) {
 	return (
 		<Layout
 			header={ <ItemHeader model={ kit } buttons={ headerButtons }/> }
-			sidebar={ <OverviewSidebar model={ kit } groupedKitContent={ documentsByType }/> }
+			sidebar={ <OverviewSidebar model={ kit } groupedKitContent={ documentsByType } isLoading={ isFetching } /> }
```
Comment (matipojo) [inline core/app/modules/kit-library/assets/js/pages/overview/overview.js:44]:
```suggestion
			sidebar={ <OverviewSidebar model={ kit } groupedKitContent={ documentsByType } isFetching ={ isFetching } /> }
```

File: core/app/modules/kit-library/data/repository.php [hunk 8]
```diff
@@ -27,43 +29,58 @@ class Repository {
...
+	 * @param      $id
+	 * @param null $user_id
+	 * @param bool $manifest_included
 	 *
-	 * @return array|null
+	 * @return array
 	 * @throws Wp_Error_Exception
 	 */
-	public function find( $id ) {
+	public function find( $id, $user_id = null, $manifest_included = true ) {
```
Comment (matipojo) [inline core/app/modules/kit-library/data/repository.php:57]:
```suggestion
	public function find( $id, $user_id = null, $options = [] ) {
```

File: core/app/modules/kit-library/data/repository.php [hunk 9]
```diff
@@ -104,6 +121,58 @@ public function get_download_link( $id ) {
...
+	 * @throws Wp_Error_Exception
+	 */
+	public function remove_from_favorites( $user_id, $id ) {
+		// To check that the kit is exists.
+		$kit = $this->find( $id, $user_id, false );
+		$favorites = $this->get_user_favorites_meta( $user_id );
+
+		if ( ! in_array( $kit['id'], $favorites, true ) ) {
+			return $kit;
+		}
```
Comment (matipojo) [inline core/app/modules/kit-library/data/repository.php:163]: Remove

File: includes/template-library/sources/local.php [hunk 10]
```diff
@@ -429,7 +429,7 @@ public function get_items( $args = [] ) {
 			foreach ( $templates_query->get_posts() as $post ) {
 				$templates[] = $this->get_item( $post->ID );
 			}
-		}
+		};
```
Comment (matipojo) [inline includes/template-library/sources/local.php:432]: Remove

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/app/modules/kit-library/data/repository.php [hunk 11]
```diff
@@ -104,6 +125,60 @@ public function get_download_link( $id ) {
...
+		$favorites = array_filter( $favorites, function ( $item ) use ( $kit ) {
+			return $item !== $kit['id'];
+		} );
+
+		$this->save_user_favorites_meta( $user_id, $favorites );
+
+		$kit['is_favorite'] = false;
+
+		return $kit;
+	}
```
Comment (matipojo) [inline core/app/modules/kit-library/data/repository.php:180]: Can you re-use the similar existing method?
https://github.com/elementor/elementor/blob/31629fa3b3b45b58356208102016a8406d21a7eb/includes/template-library/sources/base.php#L190-L204

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/app/modules/kit-library/data/kits/endpoints/favorites.php [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+	 * @param string           $id
+	 * @param \WP_REST_Request $request
+	 *
+	 * @return \WP_Error|\WP_REST_Response
+	 */
+	public function create_item( $id, $request ) {
+		$repository = $this->controller->get_repository();
+
+		try {
+			$kit = $repository->add_to_favorites( $id );
```
Comment (matipojo) [inline core/app/modules/kit-library/data/kits/endpoints/favorites.php:62]: $kit or success?

File: core/app/modules/kit-library/data/repository.php [hunk 13]
```diff
@@ -33,37 +38,54 @@ class Repository {
...
+		] );
+
 		$item = $this->get_kits_data()
 			->find( function ( $kit ) use ( $id ) {
 				return $kit->_id === $id;
 			} );
 
 		if ( ! $item ) {
-			return null;
+			throw new Wp_Error_Exception(
```
Comment (matipojo) [inline core/app/modules/kit-library/data/repository.php:69]: A `find` method should throw an error?

File: core/app/modules/kit-library/data/repository.php [hunk 14]
```diff
@@ -104,6 +126,40 @@ public function get_download_link( $id ) {
...
+	 * @param $id
+	 *
+	 * @return array
+	 * @throws Wp_Error_Exception
+	 */
+	public function add_to_favorites( $id ) {
+		// To check that the kit is exists.
+		$kit = $this->find( $id, [ 'manifest_included' => false ] );
+
+		$this->user_favorites->add( 'elementor', 'kits', $kit['id'] );
```
Comment (matipojo) [inline core/app/modules/kit-library/data/repository.php:139]:
```suggestion
		$success = $this->user_favorites->add( 'elementor', 'kits', $kit['id'] );
```

File: core/app/modules/kit-library/data/repository.php [hunk 15]
```diff
@@ -167,6 +224,7 @@ private function transform_kit_api_response( $kit, $manifest = null ) {
 				'access_level' => $kit->access_level,
 				'keywords' => $kit->keywords,
 				'taxonomies' => $taxonomies,
+				'is_favorite' => in_array( $kit->_id, $favorites, true ),
```
Comment (matipojo) [inline core/app/modules/kit-library/data/repository.php:227]:
```suggestion
				'is_favorite' => $this->is_favorite( $kit ),
```

File: modules/library/user-favorites.php [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+	 * @param array $value
+	 *
+	 * @return $this
+	 */
+	public function save( $vendor, $resource, $value = [] ) {
+		$all_favorites = $this->get();
+
+		$all_favorites[ $this->get_key( $vendor, $resource ) ] = $value;
+
+		update_user_meta( $this->user_id, self::USER_META_KEY, $all_favorites );
```
Comment (matipojo) [inline modules/library/user-favorites.php:67]:
```suggestion
		$success = update_user_meta( $this->user_id, self::USER_META_KEY, $all_favorites );
		
		if ( $success ) ...
```
Here and all methods should return a bool (success/failed)

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/library/user-favorites.php [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+	 */
+	public function save( $vendor, $resource, $value = [] ) {
+		$all_favorites = $this->get();
+
+		$all_favorites[ $this->get_key( $vendor, $resource ) ] = $value;
+
+		$result = update_user_meta( $this->user_id, self::USER_META_KEY, $all_favorites );
+
+		if ( false === $result ) {
+			throw new \Exception( 'Failed to save user favorites.' );
```
Comment (matipojo) [inline modules/library/user-favorites.php:82]: Translate

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — APPROVED** [review]

**Review by matipojo — APPROVED** [review]

**Pull request merged**